src/ast/passes/semantic_analyser.cpp

```diff
@@ -3531,35 +3576,75 @@ void SemanticAnalyser::accept_statements(StatementList &stmts)
   }
 }
 
-void SemanticAnalyser::update_key_type(const Map &map, const MapKey &new_key)
+void SemanticAnalyser::update_key_type(const Map &map,
+                                       const SizedType &new_key_type,
+                                       bool is_tuple_ast)
 {
   if (const auto &key = map_key_.find(map.ident); key != map_key_.end()) {
     bool valid = true;
-    if (key->second.args_.size() == new_key.args_.size()) {
-      for (size_t i = 0; i < key->second.args_.size(); i++) {
-        SizedType &key_type = key->second.args_[i];
-        const SizedType &new_key_type = new_key.args_[i];
-        if (key_type.IsStringTy() && new_key_type.IsStringTy()) {
-          key_type.SetSize(
-              std::max(key_type.GetSize(), new_key_type.GetSize()));
-        } else if (key_type != new_key_type) {
+    bool replace_key = false;
+    if (key->second.GetTy() == new_key_type.GetTy()) {
+      if (key->second.IsStringTy()) {
+        key->second.SetSize(
+            std::max(key->second.GetSize(), new_key_type.GetSize()));
+      } else if (key->second.IsTupleTy()) {
+        auto &keyFields = key->second.GetFields();
+        auto &newKeyFields = new_key_type.GetFields();
+        if (keyFields.size() != newKeyFields.size()) {
           valid = false;
-          break;
+        } else {
+          for (size_t i = 0; i < keyFields.size(); i++) {
+            SizedType &key_type = keyFields[i].type;
+            const SizedType &key_field_type = newKeyFields[i].type;
+            if (key_type.GetTy() != key_field_type.GetTy()) {
+              valid = false;
+              break;
+              // We can't exchange the keys if we don't have access to the
+              // underlying expressions
+            } else if (key_type.IsStringTy() && is_tuple_ast) {
+              // We have to replace the whole key type because of tuple
+              // alignment
```

Comment:
Why restrict this to literal tuples in the AST? If we relaxed this then that incompatible string-length limitation when using non-literal tuples would be gone.

Reply:
Good catch. I think this was leftover from when I was attempting to manually update individual keys. I'll remove the `is_tuple_ast` check.